<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Analyzing single cell DNA sequencing data • scDNA</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Analyzing single cell DNA sequencing data">
<meta name="description" content="This package contains functions for analyzing single cell DNA-sequencing data. Functions primarily focus on data exported from Mission Bio's Tapestri pipeline.">
<meta property="og:description" content="This package contains functions for analyzing single cell DNA-sequencing data. Functions primarily focus on data exported from Mission Bio's Tapestri pipeline.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">scDNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/CNV_example_vignette.html">CopyNumberVariation_tutorial</a></li>
    <li><a class="dropdown-item" href="articles/Demultiplex_vignette.html">scDNA: Example for demultiplexing samples</a></li>
    <li><a class="dropdown-item" href="articles/vignette.html">scDNA: R package for DNA+protein single cell DNA sequencing analysis</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="scdna-v11">scDNA v1.1<a class="anchor" aria-label="anchor" href="#scdna-v11"></a>
</h1></div>
<!-- badges: start -->
<!-- badges: end -->
<p>The goal of scDNA R package is to provide a simple framework for analyzing single cell DNA sequencing data. The current version primarily focuses processing variant information on the Mission Bio Tapestri platform. Functionality includes import of h5 files from Tapestri pipeline, basic variant annotation, genotype extraction, clone identification, and clonal trajectory inference. This package provides wrappers for normalizing protein data for scDNA+Protein libraries for downstream analysis.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install (re-install) the current version (1.1) of scDNA below This package is best suited for R &gt;=V4.3.1 and will function on standard workstations. Typical install time should be less than 5 minutes. Average run time from beginning to end of analysis will vary by sample size but should not exceed 30 minutes.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"bowmanr/scDNA"</span>,force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="version-updates">Version Updates<a class="anchor" aria-label="anchor" href="#version-updates"></a>
</h2>
<div class="section level3">
<h3 id="v11">
<strong>v1.1</strong><a class="anchor" aria-label="anchor" href="#v11"></a>
</h3>
<p>Version 1.1 is finally here with exciting new developments: - New sequencing panels for variant annotation introduced: - hg38 - mm10<br>
- New plotting functions for RL trajectories. - new interactive plots, - BSCITE-style implementation. - Demultiplexing samples is introduced - (integrated and adapted from <a href="https://www.biorxiv.org/content/10.1101/2022.09.20.508786v1.full" class="external-link">Robinson et al</a>, <a href="https://github.com/RobinsonTroy/scMRD" class="external-link">github</a>) - vignette included to demonstrate how to perform it. - Cell confidence labeling based on DNA and Protein data. - Outlier scores introduced for cell confidence. - Stain index introduced for cell confidence. - Copy number variation (CNV) and Ploidy analysis introduced. - Allele dropout assessment introduced.</p>
</div>
<div class="section level3">
<h3 id="v101">
<strong>v1.0.1</strong><a class="anchor" aria-label="anchor" href="#v101"></a>
</h3>
<ul>
<li>
<p>H5 files are now read using the rhdf5 package and stored into a <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a> container.</p>
<ul>
<li><p>Merged h5 samples are identified and sample names are stored in colData(). Variant identification is ran separately and then merged.</p></li>
<li><p>Variant information is stored in rowData()</p></li>
<li><p>NGT matrix, clonal abundance, and clone architecture <a href="https://bowmanr.github.io/scDNA_myeloid/">familiar to previous versions</a> can be found in the metadata.</p></li>
</ul>
</li>
<li>
<p>Variant identification and annotation is performed initially before reading in all the genotyping/QC data.</p>
<ul>
<li><p>Transcript annotation matches <a href="https://docs.cbioportal.org/mutation-data-transcript-annotation/#transcript-assignment" class="external-link">cannonical transcripts</a> used in the <a href="https://www.cbioportal.org/" class="external-link">cBio portal</a>.</p></li>
<li><p>To decrease variant location identification runtime, we created a custom TxDB object for the Clonal Evolution Panel from used <a href="https://www.nature.com/articles/s41586-020-2864-x" class="external-link">here</a>. If you have a different panel you can also use the <a href="https://bioconductor.org/packages/release/data/annotation/html/TxDb.Hsapiens.UCSC.hg19.knownGene.html" class="external-link">TxDB for hg19 from UCSC.</a> Future versions will have local data for all panels from Mission Bio, as well as a simple script for generating a TxDB object for custom panels.</p></li>
</ul>
</li>
<li>
<p>Protein data is stored as an altExp() container within the container.</p>
<ul>
<li><p>Wrappers for <a href="https://github.com/niaid/dsb" class="external-link">DSB</a> and CLR normalization are provided. (CLR currently performed in <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>).</p></li>
<li><p>Simple import into Seurat is demonstrated.</p></li>
<li><p>Export to FCS files with mutations and clone “completeness” provided as variables.</p></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="simple-workflow">Simple workflow<a class="anchor" aria-label="anchor" href="#simple-workflow"></a>
</h2>
<p>Identify all variants within a sample.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bowmanr.github.io/scDNA/">scDNA</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">sample_file</span><span class="op">&lt;-</span> <span class="st">"test_file.h5"</span></span>
<span><span class="va">variant_output</span><span class="op">&lt;-</span><span class="fu"><a href="reference/variant_ID.html">variant_ID</a></span><span class="op">(</span>file<span class="op">=</span><span class="va">sample_file</span>,</span>
<span>                           panel<span class="op">=</span><span class="st">"MSK_RL"</span>, <span class="co"># "UCSC" can be used for other panels</span></span>
<span>                           GT_cutoff<span class="op">=</span><span class="fl">0</span>,  <span class="co"># mimimum percent of cells where a successful genotyping call was made</span></span>
<span>                           VAF_cutoff<span class="op">=</span><span class="fl">0</span><span class="op">)</span> <span class="co"># mimimum variant allele frequency </span></span></code></pre></div>
<p>Identify mutations in genes of interest.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_of_interest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IDH2"</span>,<span class="st">"NRAS"</span>,<span class="st">"NPM1"</span>,<span class="st">"TET2"</span>,<span class="st">"FLT3"</span>,<span class="st">"IDH1"</span><span class="op">)</span></span>
<span><span class="va">variants_of_interest</span><span class="op">&lt;-</span><span class="va">variant_output</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Class</span><span class="op">==</span><span class="st">"Exon"</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">VAF</span><span class="op">&gt;</span><span class="fl">0.01</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">genotyping_rate</span><span class="op">&gt;</span><span class="fl">85</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">CONSEQUENCE</span><span class="op">)</span><span class="op">&amp;</span><span class="va">CONSEQUENCE</span><span class="op">!=</span><span class="st">"synonymous"</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">SYMBOL</span><span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span><span class="va">genes_of_interest</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>   </span>
<span>                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">VAF</span><span class="op">)</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="co"># take the 3 most abundance mutations</span></span></code></pre></div>
<p>Read in the data, enumerate clones, and compute statistics. Sample statistics mirror that seen in Figure 1 <a href="https://www.nature.com/articles/s41586-020-2864-x" class="external-link">here</a>, and are stored in the metadata.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">&lt;-</span><span class="fu"><a href="reference/tapestri_h5_to_sce.html">tapestri_h5_to_sce</a></span><span class="op">(</span>file<span class="op">=</span><span class="va">sample_file</span>,variant_set <span class="op">=</span> <span class="va">variants_of_interest</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">&lt;-</span><span class="fu"><a href="reference/enumerate_clones.html">enumerate_clones</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">&lt;-</span><span class="fu"><a href="reference/compute_clone_statistics.html">compute_clone_statistics</a></span><span class="op">(</span><span class="va">sce</span>,skip_ploidy<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Simple function for producing a graph in the style of Figure 1D from <a href="https://www.nature.com/articles/s41586-020-2864-x" class="external-link">here</a>,</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/clonograph.html">clonograph</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/Screen%20Shot%202023-09-19%20at%2010.07.16%20PM.png" width="373"></p>
<p>Function to perform Reinforcment Learning / MDP approach for clonal trajectory as in Figure 3 <a href="https://www.nature.com/articles/s41586-020-2864-x" class="external-link">here</a>,</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">&lt;-</span><span class="fu"><a href="reference/trajectory_analysis.html">trajectory_analysis</a></span><span class="op">(</span><span class="va">sce</span>,use_ADO<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Methods for protein normalization. Both dsb and CLR normalization can be performed and stored in separate slots. We tend to have favor dsb so far.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">droplet_metadata</span><span class="op">&lt;-</span> <span class="fu"><a href="reference/extract_droplet_size.html">extract_droplet_size</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">background_droplets</span><span class="op">&lt;-</span><span class="va">droplet_metadata</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Droplet_type</span><span class="op">==</span><span class="st">"Empty"</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">dna_size</span><span class="op">&lt;</span><span class="fl">1.5</span><span class="op">&amp;</span><span class="va">dna_size</span><span class="op">&gt;</span><span class="fl">0.15</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                          <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">Cell</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span><span class="op">&lt;-</span><span class="fu"><a href="reference/normalize_protein_data.html">normalize_protein_data</a></span><span class="op">(</span>sce<span class="op">=</span><span class="va">sce</span>,</span>
<span>                             metadata<span class="op">=</span><span class="va">droplet_metadata</span>,</span>
<span>                             method<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dsb"</span>,<span class="st">"CLR"</span><span class="op">)</span>,</span>
<span>                             detect_IgG<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                             background_droplets<span class="op">=</span><span class="va">background_droplets</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="developments-in-progress">Developments in progress:<a class="anchor" aria-label="anchor" href="#developments-in-progress"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Cohort summarization</li>
<li>Creating custom TxDB objects</li>
</ol>
</div>
<div class="section level3">
<h3 id="ongoing-investigation">Ongoing investigation:<a class="anchor" aria-label="anchor" href="#ongoing-investigation"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Improving cell identification and distinction from empty droplets.
<ol style="list-style-type: decimal">
<li>Doublet and dead cell identification</li>
</ol>
</li>
<li>Improve normalization for protein data.
<ol style="list-style-type: decimal">
<li>Improve cell type identification based on immunophenotype</li>
</ol>
</li>
<li>Improvements to the MDP and RL.</li>
</ol>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing scDNA</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Bobby Bowman <br><small class="roles"> Author, maintainer </small>   </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bobby Bowman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
