---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# scDNA

<!-- badges: start -->
<!-- badges: end -->

The goal of scDNA R package is for DNA+protein single cell DNA sequencing analysis. This includes extracting data from Tapestri, to produce genotype matrix (NGT), determine variant allele frequency(VAF), summarize clone information, and determine the clonal trajectories.
## Installation

You can install the development version of scDNA like so:

```{r}
devtools::install_github("bowmanr/scDNA")
```
## Example

This is a basic example which shows you how to solve a common problem:

```{r example}
library(scDNA)
## basic example code
file<-"~/path_to_my_h5/Sample.dna+protein.h5"
NGT<-read_tapestri_h5_NGT(file,
                          VAF_cutoff = 0.05,
                          GT_cutoff = 10)
filtered_NGT<-quality_filter_NGT(file=file,
                                 NGT=NGT,
                                 DP_cut=10, 
                                 AF_cut=20,
                                 GQ_cut=20)
print(dim(NGT))
print(dim(filtered_NGT))
#Total number of NAs
print(table(is.na(NGT)))
print(table(is.na(filtered_NGT)))
```

