<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Annotate variants of interest This function takes in h5 files to extract DNA information through given TXDB files (primarily hg19) The gene names are mapped for their specific nucleotide positions. The starting and ending positions for the chromosome are listed. The consequence of the mutation such as synonymous, nonsynonymous, as well as if this is a coding region, on the exon boundry or intronic. Short amino acid changes are also labeled. The variance matrix also extracts the amplicon that the variant is found on. — annotate_variants • scDNA</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Annotate variants of interest This function takes in h5 files to extract DNA information through given TXDB files (primarily hg19) The gene names are mapped for their specific nucleotide positions. The starting and ending positions for the chromosome are listed. The consequence of the mutation such as synonymous, nonsynonymous, as well as if this is a coding region, on the exon boundry or intronic. Short amino acid changes are also labeled. The variance matrix also extracts the amplicon that the variant is found on. — annotate_variants"><meta name="description" content="Annotate variants of interest
This function takes in h5 files to extract DNA information through given TXDB files (primarily hg19)
The gene names are mapped for their specific nucleotide positions. The starting and ending positions
for the chromosome are listed. The consequence of the mutation such as synonymous, nonsynonymous,
as well as if this is a coding region, on the exon boundry or intronic. Short amino acid changes are also labeled.
The variance matrix also extracts the amplicon that the variant is found on."><meta property="og:description" content="Annotate variants of interest
This function takes in h5 files to extract DNA information through given TXDB files (primarily hg19)
The gene names are mapped for their specific nucleotide positions. The starting and ending positions
for the chromosome are listed. The consequence of the mutation such as synonymous, nonsynonymous,
as well as if this is a coding region, on the exon boundry or intronic. Short amino acid changes are also labeled.
The variance matrix also extracts the amplicon that the variant is found on."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scDNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Demultiplex_vignette.html">scDNA: Example for demultiplexing samples</a></li>
    <li><a class="dropdown-item" href="../articles/vignette.html">scDNA: R package for DNA+protein single cell DNA sequencing analysis</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Annotate variants of interest This function takes in h5 files to extract DNA information through given TXDB files (primarily hg19) The gene names are mapped for their specific nucleotide positions. The starting and ending positions for the chromosome are listed. The consequence of the mutation such as synonymous, nonsynonymous, as well as if this is a coding region, on the exon boundry or intronic. Short amino acid changes are also labeled. The variance matrix also extracts the amplicon that the variant is found on.</h1>

      <div class="d-none name"><code>annotate_variants.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Annotate variants of interest
This function takes in h5 files to extract DNA information through given TXDB files (primarily hg19)
The gene names are mapped for their specific nucleotide positions. The starting and ending positions
for the chromosome are listed. The consequence of the mutation such as synonymous, nonsynonymous,
as well as if this is a coding region, on the exon boundry or intronic. Short amino acid changes are also labeled.
The variance matrix also extracts the amplicon that the variant is found on.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">annotate_variants</span><span class="op">(</span><span class="va">file</span>, panel <span class="op">=</span> <span class="cn">NULL</span>, select_variants <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-file">file<a class="anchor" aria-label="anchor" href="#arg-file"></a></dt>
<dd><p>path to h5 file to pull out relevant DNA information</p></dd>


<dt id="arg-select-variants">select_variants<a class="anchor" aria-label="anchor" href="#arg-select-variants"></a></dt>
<dd><p>variants of interest, default is all variants</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>variant annotation matrix</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bobby Bowman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

