<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ scDNA</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scDNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Demultiplex_vignette.html">scDNA: Example for demultiplexing samples</a></li>
    <li><a class="dropdown-item" href="../articles/vignette.html">scDNA: R package for DNA+protein single cell DNA sequencing analysis</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="BuildMDP.html">BuildMDP()</a></code>

        </dt>
        <dd>The main function builds an adjacency list of the theoretical order of mutations. The same MDP mask can be reused different mutations</dd>
      </dl><dl><dt>

          <code><a href="GetMembership.html">GetMembership()</a></code>

        </dt>
        <dd>Get membership quality of cells vs empty</dd>
      </dl><dl><dt>

          <code><a href="GetStainIndex.html">GetStainIndex()</a></code>

        </dt>
        <dd>Get Stain Index for cells</dd>
      </dl><dl><dt>

          <code><a href="add_cell_annotation.html">add_cell_annotation()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="annotate_variants.html">annotate_variants()</a></code>

        </dt>
        <dd>Annotate variants of interest This function takes in h5 files to extract DNA information through given TXDB files (primarily hg19) The gene names are mapped for their specific nucleotide positions. The starting and ending positions for the chromosome are listed. The consequence of the mutation such as synonymous, nonsynonymous, as well as if this is a coding region, on the exon boundry or intronic. Short amino acid changes are also labeled. The variance matrix also extracts the amplicon that the variant is found on.</dd>
      </dl><dl><dt>

          <code><a href="attach_weights.html">attach_weights()</a></code>

        </dt>
        <dd>Assigns observed counts as weights to the Markov Decision Process</dd>
      </dl><dl><dt>

          <code><a href="cell_confidence_labeling.html">cell_confidence_labeling()</a></code>

        </dt>
        <dd>cell_confidence_labeling This function provides additional labeling to he quality of the cells based on dna and protein data.</dd>
      </dl><dl><dt>

          <code><a href="clonograph.html">clonograph()</a></code>

        </dt>
        <dd>Plotting clonographs</dd>
      </dl><dl><dt>

          <code><a href="compare_VAFs.html">compare_VAFs()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="compute_clone_statistics.html">compute_clone_statistics()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="demultiplex_samples.html">demultiplex_samples()</a></code>

        </dt>
        <dd>Demultiplex wrapper function to handle splitting samples</dd>
      </dl><dl><dt>

          <code><a href="enumerate_clones.html">enumerate_clones()</a></code>

        </dt>
        <dd>Enumerate clones</dd>
      </dl><dl><dt>

          <code><a href="extract_droplet_size.html">extract_droplet_size()</a></code>

        </dt>
        <dd>Extract protein library size, dna library size, and amplicon size for all droplets.</dd>
      </dl><dl><dt>

          <code><a href="fcs_export.html">fcs_export()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="gemerate_txdb.html">gemerate_txdb()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="get_own_path.html">get_own_path()</a></code>

        </dt>
        <dd>A way to navigate the MDP for any given starting root node to leaf node.</dd>
      </dl><dl><dt>

          <code><a href="impute_cluster.html">impute_cluster()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="loom_to_sce.html">loom_to_sce()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="match_clonal_graph.html">match_clonal_graph()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="mdp_Q_learning_with_linklist.html">mdp_Q_learning_with_linklist()</a></code>

        </dt>
        <dd>This file run Reinforcement Learning (model-free Q-learning) to evaluate the most likely mutation paths from the data</dd>
      </dl><dl><dt>

          <code><a href="normalize_protein_data.html">normalize_protein_data()</a></code>

        </dt>
        <dd>Normalize Protein Data</dd>
      </dl><dl><dt>

          <code><a href="optimize_matrix.html">optimize_matrix()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="quality_output.html">quality_output()</a></code>

        </dt>
        <dd>Produce long form quality metrics for each cell-variant pair for read depth, allele frequency, and genotype quality</dd>
      </dl><dl><dt>

          <code><a href="readDNA_CN_H5.html">readDNA_CN_H5()</a></code>

        </dt>
        <dd>This function generates the Copy Number by determining the ploidy of each mutation</dd>
      </dl><dl><dt>

          <code><a href="select_clones.html">select_clones()</a></code>

        </dt>
        <dd>Select clones of interest on the basis QC metrics</dd>
      </dl><dl><dt>

          <code><a href="tabulate_mutations.html">tabulate_mutations()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="tapestri_h5_to_sce.html">tapestri_h5_to_sce()</a></code>

        </dt>
        <dd>Import Tapestri H5 data and extract genotype matrix</dd>
      </dl><dl><dt>

          <code><a href="trajectory_analysis.html">trajectory_analysis()</a></code>

        </dt>
        <dd>Run Trajectory Analysis after extraction from SingleCellExperiment object</dd>
      </dl><dl><dt>

          <code><a href="trajectory_of_interest_BSCITE_format.html">trajectory_of_interest_BSCITE_format()</a></code>

        </dt>
        <dd>BSCITE Style single trajectory Plot</dd>
      </dl><dl><dt>

          <code><a href="trajectory_of_interest_figure.html">trajectory_of_interest_figure()</a></code>

        </dt>
        <dd>Individual trajectory of interest Plots</dd>
      </dl><dl><dt>

          <code><a href="variant_ID.html">variant_ID()</a></code>

        </dt>
        <dd>Variant identification and frequency tallies The DNA variants for each cell are pulled from the specified H5 files. Each variant for each cell is genotyped to be WildType, Heterozygous, Homozygous or Missing. The genotyping rate is determined by taking WT+Het+Hom over total cell calls (including missing). The VAF is determined by number of allele copies we see in a weighted sum. A filter is applied to both of these calculatins to include or exclude variants of interest. These are then annotated to include the variant information such as gene name, nucleotide location, and short amino acid changes.</dd>
      </dl><dl><dt>

          <code><a href="visualize_WT_dominant_clone.html">visualize_WT_dominant_clone()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="visualize_all_WT_dominant_clone.html">visualize_all_WT_dominant_clone()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="visualize_any_optimal_path.html">visualize_any_optimal_path()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="visualize_full_network.html">visualize_full_network()</a></code>

        </dt>
        <dd>Visualize Full Network</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bobby Bowman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

