<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Produce long form quality metrics for each cell-variant pair for read depth, allele frequency, and genotype quality — quality_output • scDNA</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Produce long form quality metrics for each cell-variant pair for read depth, allele frequency, and genotype quality — quality_output"><meta name="description" content="Produce long form quality metrics for each cell-variant pair for read depth, allele frequency, and genotype quality"><meta property="og:description" content="Produce long form quality metrics for each cell-variant pair for read depth, allele frequency, and genotype quality"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scDNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Demultiplex_vignette.html">scDNA: Example for demultiplexing samples</a></li>
    <li><a class="dropdown-item" href="../articles/vignette.html">scDNA: R package for DNA+protein single cell DNA sequencing analysis</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Produce long form quality metrics for each cell-variant pair for read depth, allele frequency, and genotype quality</h1>

      <div class="d-none name"><code>quality_output.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Produce long form quality metrics for each cell-variant pair for read depth, allele frequency, and genotype quality</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">quality_output</span><span class="op">(</span></span>
<span>  <span class="va">file</span>,</span>
<span>  filter <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">input_variants</span>,</span>
<span>  <span class="va">input_cells</span>,</span>
<span>  <span class="va">NGT</span>,</span>
<span>  DP_cut <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  AF_cut <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  GQ_cut <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-file">file<a class="anchor" aria-label="anchor" href="#arg-file"></a></dt>
<dd><p>path to H5 file</p></dd>


<dt id="arg-filter">filter<a class="anchor" aria-label="anchor" href="#arg-filter"></a></dt>
<dd><p>default=TRUE, determines whether function should output filtered or unfiltered long form matrix containing per cell-variant pair of NGT, DP, AF, GQ metrics</p></dd>


<dt id="arg-input-variants">input_variants<a class="anchor" aria-label="anchor" href="#arg-input-variants"></a></dt>
<dd><p>variants of interest to collect QC data</p></dd>


<dt id="arg-input-cells">input_cells<a class="anchor" aria-label="anchor" href="#arg-input-cells"></a></dt>
<dd><p>cells of interest to subset on</p></dd>


<dt id="arg-ngt">NGT<a class="anchor" aria-label="anchor" href="#arg-ngt"></a></dt>
<dd><p>NGT matrix</p></dd>


<dt id="arg-dp-cut">DP_cut<a class="anchor" aria-label="anchor" href="#arg-dp-cut"></a></dt>
<dd><p>read depth cutoff</p></dd>


<dt id="arg-af-cut">AF_cut<a class="anchor" aria-label="anchor" href="#arg-af-cut"></a></dt>
<dd><p>allele frequency cutoff</p></dd>


<dt id="arg-gq-cut">GQ_cut<a class="anchor" aria-label="anchor" href="#arg-gq-cut"></a></dt>
<dd><p>genotype quality cutoff</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>filtered NGT matrix containing NAs for cells that do not pass QC</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bobby Bowman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

