<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>scDNA: Example for demultiplexing samples â€¢ scDNA</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="scDNA: Example for demultiplexing samples">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scDNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Cell_type_VAFs.html">Cell Type VAF vignette</a></li>
    <li><a class="dropdown-item" href="../articles/CNV_example_vignette.html">CopyNumberVariation_tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/Demultiplex_vignette.html">scDNA: Example for demultiplexing samples</a></li>
    <li><a class="dropdown-item" href="../articles/vignette.html">scDNA: R package for DNA+protein single cell DNA sequencing analysis</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>scDNA: Example for demultiplexing samples</h1>
                        <h4 data-toc-skip class="author">Michael Bowman
and Bobby Bowman</h4>
            
            <h4 data-toc-skip class="date">2024-07-01</h4>
      

      <div class="d-none name"><code>Demultiplex_vignette.Rmd</code></div>
    </div>

    
    
<p>install our package if you have not done so already.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://remotes.r-lib.org" class="external-link">remotes</a></span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"bowmanr/scDNA@dev"</span>,force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Load in libraries</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bowmanr.github.io/scDNA/">scDNA</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tkonopka/umap" class="external-link">umap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://kwstat.github.io/pals/" class="external-link">pals</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">superheat</span><span class="op">)</span></span></code></pre></div>
<p>We assume you have already familiarized yourself with our single
sample vignette, here we should the demultiplexing sample
possibilities.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file_bbmix</span><span class="op">=</span><span class="st">"~/Lab/Core/scDNA_paper_figures/H5/BB_Mix1.dna+protein.h5"</span></span></code></pre></div>
<p>run variant ID like normal, with no additional inputs, BUT we choose
extremely high VAF and Genotyping Cut offs to detect SNPs.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variant_output</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/variant_ID.html">variant_ID</a></span><span class="op">(</span><span class="va">file_bbmix</span>,</span>
<span>                            GT_cutoff<span class="op">=</span><span class="fl">50</span>,</span>
<span>                             VAF_cutoff<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>Filter variants of interest to reduce the SNPs we need.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variant_set</span><span class="op">&lt;-</span><span class="va">variant_output</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span> <span class="co"># filter(Het&lt;Hom|Het&lt;WT)%&gt;%</span></span>
<span> <span class="co"># filter(WT&gt;0&amp;Hom&gt;0)%&gt;%</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">genotyping_rate</span><span class="op">&gt;</span><span class="fl">90</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">VAF</span><span class="op">&gt;</span><span class="fl">1</span><span class="op">&amp;</span><span class="va">VAF</span><span class="op">&lt;</span><span class="fl">99</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">final_annot</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>final_annot <span class="op">=</span> <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span> <span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">final_annot</span>,<span class="st">"_"</span>,<span class="va">LETTERS</span><span class="op">[</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/row_number.html" class="external-link">row_number</a></span><span class="op">(</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">}</span> </span>
<span>                             <span class="kw">else</span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">final_annot</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>run the tapestri thing like normal</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">&lt;-</span><span class="fu">scDNA</span><span class="fu">::</span><span class="fu"><a href="../reference/tapestri_h5_to_sce.html">tapestri_h5_to_sce</a></span><span class="op">(</span>file<span class="op">=</span><span class="va">file_bbmix</span>,</span>
<span>                    variant_set <span class="op">=</span> <span class="va">variant_set</span><span class="op">)</span></span></code></pre></div>
<p>Preform demultiplex to get cluster assignment for each cell. We use
an iterative learning strategy to reduce our variants by cells matrix to
full rank (as dense as possible!). The sensitivity_threshold determines
how dense we are looking for our sample. The first value impacts
variants to keep and ditch, while the second values kicks out poorly
genotyped cells. Generally, the cells threshold should be lower than the
varaints one. Lastly, expected_samples is how many samples we are
expected to demultiplex.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">cell_clust_df</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/demultiplex_samples.html">demultiplex_samples</a></span><span class="op">(</span><span class="va">sce</span>,sensitivity_threshold<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.01</span>,<span class="fl">0.0001</span><span class="op">)</span>,expected_samples <span class="op">=</span><span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>Now perform we perform variant_ID but pump in the cell-cluster
assignment dataframe. This splits each sample, as if it is a cohort.
Afterwards we can loop our variants of interest for each specific
sample. We then supply the cells of interest along with the variants of
interest into tapestri_h5_to_sce to obtain a list of sce objects. The
example below uses the same filtering for each sample for brevity.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"~/Lab/Core/scDNA-dev/R/variant_ID.R"</span><span class="op">)</span></span>
<span><span class="va">variant_output_true</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/variant_ID.html">variant_ID</a></span><span class="op">(</span><span class="va">file_bbmix</span>,panel<span class="op">=</span><span class="st">"MSK_RL"</span>,demultiplexed <span class="op">=</span> <span class="va">cell_clust_df</span><span class="op">)</span></span>
<span><span class="va">genes_of_interest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IDH2"</span>,<span class="st">"NRAS"</span>,<span class="st">"NPM1"</span>,<span class="st">"TET2"</span>,<span class="st">"FLT3"</span>,<span class="st">"IDH1"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce_list</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">sample_iter</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">variant_output_true</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">variant_output_specific</span><span class="op">&lt;-</span><span class="va">variant_output_true</span><span class="op">[[</span><span class="va">sample_iter</span><span class="op">]</span><span class="op">]</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Class</span><span class="op">==</span><span class="st">"Exon"</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">CONSEQUENCE</span><span class="op">)</span><span class="op">&amp;</span><span class="va">CONSEQUENCE</span><span class="op">!=</span><span class="st">"synonymous"</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">VAF</span><span class="op">&gt;</span><span class="fl">1</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">genotyping_rate</span><span class="op">&gt;</span><span class="fl">85</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">SYMBOL</span><span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span><span class="va">genes_of_interest</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>   </span>
<span>                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">VAF</span><span class="op">)</span><span class="op">)</span><span class="co">#%&gt;%</span></span>
<span></span>
<span>  <span class="va">sce_list</span><span class="op">[</span><span class="va">sample_iter</span><span class="op">]</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/tapestri_h5_to_sce.html">tapestri_h5_to_sce</a></span><span class="op">(</span>file<span class="op">=</span><span class="va">file_bbmix</span>,</span>
<span>                      variant_set <span class="op">=</span> <span class="va">variant_output_specific</span>,</span>
<span>                      GT_cutoff<span class="op">=</span><span class="fl">90</span>, </span>
<span>                      VAF_cutoff<span class="op">=</span><span class="fl">0.01</span>,</span>
<span>                      DP_cutoff<span class="op">=</span><span class="fl">10</span>,</span>
<span>                      GQ_cutoff<span class="op">=</span><span class="fl">20</span>,</span>
<span>                      AF_cutoff<span class="op">=</span><span class="fl">20</span>,</span>
<span>                      demultiplex_cells<span class="op">=</span><span class="va">cell_clust_df</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                        <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">final_cluster</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">final_cluster</span><span class="op">)</span><span class="op">[</span><span class="va">sample_iter</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">cell_names</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Running a cohort analysis on a demultiplexed sample follows our base
vignette, but perform it inside a loop! OR you can save the sce_list as
independent RDS files and run the base_line vignette independently. When
integrating multiple samples together with Seurat we have found the
Harmony integration layer has worked well and recommend it</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span><span class="op">(</span><span class="va">sce_iter</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">sce_list</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">current_sce</span><span class="op">&lt;-</span><span class="va">sce_list</span><span class="op">[[</span><span class="va">sce_iter</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">current_sce</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/enumerate_clones.html">enumerate_clones</a></span><span class="op">(</span><span class="va">current_sce</span>, replicates <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span>  <span class="va">current_sce</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/compute_clone_statistics.html">compute_clone_statistics</a></span><span class="op">(</span><span class="va">current_sce</span>,skip_ploidy<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="../reference/clonograph.html">clonograph</a></span><span class="op">(</span><span class="va">current_sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bobby Bowman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
